yarn-project/end-to-end/src/integration_archiver_l1_to_l2.test.ts

```diff
@@ -0,0 +1,166 @@
+import { DebugLogger } from '@aztec/foundation/log';
+import { PortalERC20Abi, PortalERC20Bytecode, TokenPortalAbi, TokenPortalBytecode } from '@aztec/l1-artifacts';
+import { Chain, HttpTransport, PublicClient, getContract } from 'viem';
+import { delay, deployNonNativeL2TokenContract, pointToPublicKey, setNextBlockTimestamp, setup } from './utils.js';
+import { Archiver } from '@aztec/archiver';
+
+describe('e2e_l1_to_l2_msg', () => {
+  let aztecNode: AztecNodeService;
+  let aztecRpcServer: AztecRPCServer;
+  let archiver: Archiver;
+  let accounts: AztecAddress[];
+  let logger: DebugLogger;
+  let config: AztecNodeConfig;
+
+  let l2Contract: Contract;
+  let ethAccount: EthAddress;
+
+  let tokenPortalAddress: EthAddress;
+  let underlyingERC20Address: EthAddress;
+  let rollupRegistryAddress: EthAddress;
+  let tokenPortal: any;
+  let underlyingERC20: any;
+  let publicClient: PublicClient<HttpTransport, Chain>;
+
+  beforeEach(async () => {
+    let deployL1ContractsValues: DeployL1Contracts | undefined;
+    ({ aztecNode, aztecRpcServer, deployL1ContractsValues, accounts, config, logger } = await setup(2));
+    archiver = await Archiver.createAndSync(config);
+    rollupRegistryAddress = deployL1ContractsValues!.registryAddress;
+
+    const walletClient = deployL1ContractsValues.walletClient;
+    publicClient = deployL1ContractsValues.publicClient;
+
+    ethAccount = EthAddress.fromString((await walletClient.getAddresses())[0]);
+
+    // Deploy portal contracts
+    underlyingERC20Address = await deployL1Contract(walletClient, publicClient, PortalERC20Abi, PortalERC20Bytecode);
+    tokenPortalAddress = await deployL1Contract(walletClient, publicClient, TokenPortalAbi, TokenPortalBytecode);
+    underlyingERC20 = getContract({
+      address: underlyingERC20Address.toString(),
+      abi: PortalERC20Abi,
+      walletClient,
+      publicClient,
+    });
+    tokenPortal = getContract({
+      address: tokenPortalAddress.toString(),
+      abi: TokenPortalAbi,
+      walletClient,
+      publicClient,
+    });
+  }, 30_000);
+
+  afterEach(async () => {
+    await archiver.stop();
+    await aztecNode?.stop();
+    await aztecRpcServer?.stop();
+  });
+
+  const expectBalance = async (owner: AztecAddress, expectedBalance: bigint) => {
+    const ownerPublicKey = await aztecRpcServer.getAccountPublicKey(owner);
+    const [balance] = await l2Contract.methods.getBalance(pointToPublicKey(ownerPublicKey)).view({ from: owner });
+    logger(`Account ${owner} balance: ${balance}`);
+    expect(balance).toBe(expectedBalance);
+  };
+
+  it('cancelled l1 to l2 messages cannot be consumed by archiver', async () => {
+    // first initialise the portal, create a message, then cancel it
+    const initialBalance = 10n;
+    const [ownerAddress] = accounts;
+    const ownerPub = await aztecRpcServer.getAccountPublicKey(ownerAddress);
+    logger(`Deploying L2 Token contract...`);
+    l2Contract = await deployNonNativeL2TokenContract(
```

Comment:
Deploying the l2 contract should probably be part of the before rather than inside the test. Also, as it is a "non-native" it will rely on the L1 contract, which are for both the tests where this is used also deployed freshly. As these are already tightly coupled, might as well have a `deployNonNativeL2TokenContracts` that deploys all of them.

Reply:
good idea